### PR TITLE
GH-262: In a batch only commit highest offset for each partition  Fix…

### PR DESCRIPTION
**Pull request by ltalhouarne**
…es GH-262 (https://github.com/spring-projects/spring-kafka/issues/262)  Committing each offset in a batch hinders the performance. Only the highest offset for a partition should be committed from a batch.

### CODE REVIEW

File: spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java [hunk 1]
```diff
@@ -729,8 +729,13 @@ public void testBatchListenerManual() throws Exception {
 				return invocation.callRealMethod();
 			}
 			finally {
+				boolean smallOffsetCommitted = false;
 				for (Entry<TopicPartition, OffsetAndMetadata> entry : map.entrySet()) {
-					if (entry.getValue().offset() == 2) {
+					if (entry.getValue().offset() == 1) {
+						smallOffsetCommitted = true;
+						logger.error("The highest offset should be the only one committed.");
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java]: Well, it's really an `IllegalState` to be here.
According our new logic we we should never get lower offset for the `commit()`.
I'll fix on merge

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/1f1a54434fee5afc37ca8fc99cd3cc43d8f9413a and cherry-picked to `master`.

@ltalhouarne ,
Thank you for contribution.
Looking forward for more!

**Comment by garyrussell** [comment]
Hmmm - sorry; I wanted to discuss this before merging.

It really should be consistent with what we do with non-manual acks - see `updatePendingOffsets()` and `addOffset()` which does assume the highest offset is last.

**Comment by artembilan** [comment]
> I wanted to discuss this before merging.

Well, sorry to be in a rush, but I still think that we can discuss and provide another PR if that is needed.

The code is pretty simple to share here directly:
```
private void updatePendingOffsets() {
	ConsumerRecord<K, V> record = this.acks.poll();
	while (record != null) {
		addOffset(record);
		record = this.acks.poll();
	}
}

private void addOffset(ConsumerRecord<K, V> record) {
	if (!this.offsets.containsKey(record.topic())) {
		this.offsets.put(record.topic(), new HashMap<Integer, Long>());
	}
	this.offsets.get(record.topic()).put(record.partition(), record.offset());
}
...
Map<TopicPartition, OffsetAndMetadata> commits = new HashMap<>();
for (Entry<String, Map<Integer, Long>> entry : this.offsets.entrySet()) {
	for (Entry<Integer, Long> offset : entry.getValue().entrySet()) {
		commits.put(new TopicPartition(entry.getKey(), offset.getKey()),
				new OffsetAndMetadata(offset.getValue() + 1));
	}
}
```
So, we just re-map `acks` into `offsets`. That isn't obvious for me that highest is the last...

What do you mean, please and what is your suggestion for fixing the MANUAL mode?
Thanks

**Comment by garyrussell** [comment]
>What do you mean, please and what is your suggestion for fixing the MANUAL mode?

I am just saying we should be consistent in the two places where we determine the actual offset to commit (manual and auto).

I think it's reasonable to assume the last is the highest but I don't mind the extra check.

**Comment by artembilan** [comment]
Yeah... That's true, but we still don't have choice unless iterate batch to determine that highest offset for each partition in the batch.

Or is your idea that we should apply the same "extract highest offset" algorithm into the  non-manual acks for BATCH mode ?

**Comment by garyrussell** [comment]
`addOffset()` already does that, but it assumes the last is the highest.

By consistent, I mean... consistent.

One inconsistency is we're doing the map/reduce in the container-ack case after loading up the acks whereas in the manual case, we're doing it before writing to acks; that's better because the reduce is done earlier.

I am just saying we should consider doing the same for the container-ack case - i.e. instead of loading up `acks` and then map/reduce, just add the ack to `this.offsets` (if it's higher).

In other words, use the same algo regardless of manual/auto.

**Comment by artembilan** [comment]
Now I got it!

Thank you, Gary, for your patience!

@ltalhouarne , WDYT think about one more contribution?

**Comment by ltalhouarne** [comment]
@artembilan I am fine with the aforementioned approach and would be happy to contribute. Lookout for another PR addressing the new approach.

**Comment by ltalhouarne** [comment]
@artembilan @garyrussell After digging into the flow and moving the logic to the `addOffset` method, I realized that the new approach would only work with `AckMode.MANUAL`.   

`AckMode.MANUAL_IMMEDIATE` bypasses the `addOffset` method and commits for each record: 

```
if (ListenerConsumer.this.isManualImmediateAck) {
	try {
		ackImmediate(record);
	} catch (WakeupException e) {
		// ignore - not polling
	}
}
else {
	addOffset(record);
}
```

So I believe there is indeed a need for preventing all offsets in a partition to make it to `this.acks` as previously designed. 

Here is a diff of the work with the same test `testBatchListenerManual` failing:

https://github.com/spring-projects/spring-kafka/compare/1.1.x...ltalhouarne:hotfix/highest-offset-different-approach?expand=1

**Comment by ltalhouarne** [comment]
@artembilan @garyrussell,

I have created the following PR: https://github.com/spring-projects/spring-kafka/pull/267

The work in the PR adds the highest offset per partition logic to non-manual ack batch listeners before adding to `acks`. I also took the liberty of adding the logic to the `addOffset` method to avoid any assumptions as far as the last record having the highest offset.
Finally I fixed an existing test that **would not** throw an exception if a smaller offset was committed.
